Add product categories datastore

jacob-sewell opened the pull request:
Fixes #33464.
Fixes #33465.
Fixes #33466.

### All Submissions:

-   [x] Have you followed the [WooCommerce Contributing guideline](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md)?
-   [x] Does your code follow the [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/)?
-   [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change?

<!-- Mark completed items with an [x] -->

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

### Changes proposed in this Pull Request:

Creates a data store with selectors and dispatchers for manipulating Product Categories through the WooCommerce API.

### How to test the changes in this Pull Request:

All these steps are console commands to be run from the WooCommerce home page after completing initial setup.

1. Create some product categories, e.g. `wp.data.dispatch('experimental/wc/admin/products/categories').createProductCategory({ name: 'Test 1', slug: 'test1' })`
2. Test the bulk fetch selector (you will need to run this twice): `console.table(wp.data.select('experimental/wc/admin/products/categories').getProductCategories({}))`
3. Test the single selector against one of the ids you saw in the bulk fetch results: `wp.data.select('experimental/wc/admin/products/categories').getProductCategory(18)`
4. Test the update action, e.g.: `wp.data.dispatch('experimental/wc/admin/products/categories').updateProductCategory(18,{ menu_order: 2 })`
5. Rerun bulk and single fetches to verify the update happened.
6. Test the delete action, e.g.: `wp.data.dispatch('experimental/wc/admin/products/categories').deleteProductCategory(18)`
7. Rerun bulk and single fetches to verify the delete happened.

#### NOTE:

The product category API spec contains an `image` property which has its own [schema and rules](https://woocommerce.github.io/woocommerce-rest-api-docs/#product-categories), but beyond defining them in data/src/product-categories/types.ts no attempt has been made to enforce them, nor to support creating a category with an image or updating a product category image.

### Other information:

-   [x] Have you added an explanation of what your changes do and why you'd like us to include them?
-   [x] Have you written new tests for your changes, as applicable?
-   [x] Have you successfully run tests with your changes locally?
-   [x] Have you created a changelog file for each project being changed, ie `pnpm changelog add --filter=<project>`?

<!-- Mark completed items with an [x] -->

### FOR PR REVIEWER ONLY:

-   [ ] I have reviewed that everything is sanitized/escaped appropriately for any SQL or XSS injection possibilities. I made sure Linting is not ignored or disabled.

jacob-sewell committed:
First draft of product categories datastore changes.
Heavily based on product shipping classes code.

jacob-sewell committed:
First draft of product categories datastore changes.
Heavily based on product shipping classes code.

kirtangajjar committed:
Fix type safety issues in product price (#33926)

chihsuan committed:
Wrap default payment gateway strings in __() function call (#33987)

* Wrap default payment gateway strings in __() function call

* Add changelog

Commit:
Make sure get_current_screen() is available (#33990)

* Make sure get_current_screen is available

* Add changelog

roykho committed:
Add changelog validation to pre push (#33976)

Luc45 committed:
Retry login on e2e (#33997)

* Retry login on e2e

* Add changelog

Luc45 committed:
Exclude e2e folder from build (#33934)

roykho committed:
Add cache priming to trunk (#33981)

jacob-sewell committed:
Fix constant name in product categories data store index.

octaedro committed:
Product creation experience: add more context about product data (#33755)

* Add product types tooltip

* Add target `_blank`

* Add changelog

* Fix typo

* Add onChange method

* Fixed product type tooltip

* Fix icon style

Co-authored-by: Fernando Marichal <contacto@fernandomarichal.com>

joelclimbsthings committed:
Adding additional instructions around product image upload modal (#33660)

jacob-sewell committed:
Create changelog for 33464/33984.

Luc45 committed:
Add detailed logs for e2e setup/teardown (#34004)

psealock committed:
Delete changelog files based on PR 34008 (#34009)

Delete changelog files for 33934

chihsuan committed:
Fix free features is still rendered when no recommendation (#33923)

* Fix free features is still rendered when there is no recommendation

* Add changelog

* Update BusinessDetails tab to go to next step if no installable extensions

* Update spinner

samueljseay committed:
Generate a JSON file with changes from code-analyzer (#33955)

jacob-sewell committed:
Move changelog to correct folder.

joshuatf committed:
Add data store attribute terms (#33721)

* Add product attribute terms store

* Add ability to add params to URL

* Allow IdQuery for all crud functions

* Add tests around item parent keys

* Add tests around new utils

* Throw error when not all params are replaced in REST URL

* Add require attributes for terms

* Allow urlParameters to be specified

* Use namespace to detect url parameters

* Fix up selector return after applying namespace

* Clean queries to prevent sending URL param data to endpoint

* Add tests around new utils

* Remove unused method import

* Remove urlParameters argument no longer being used

* Add changelog entry

joshuatf commented on packages/js/data/src/product-categories/types.ts:
```diff
@@ -0,0 +1,61 @@
+	count: number;
+};
+
+type Query = BaseQueryParams< keyof ProductCategory > & {
```
I missed this in our previous PRs (cc @louwie17) but this will need to be a partial as most `Query` params are not required.  Otherwise all these fields will be required when querying `getProductCategories()`.

jacob-sewell replied:
I was definitely able to call `getProductCategories({})` and `getProductCategories({ slug: 'test1' })`. Is that unexpected behavior?

joshuatf replied:
Were you calling it in the console or in your editor?  Typescript is not available at runtime and should only show errors or hinting in your editor during development.

jacob-sewell replied:
Yeah, this is exactly what was going on. I feel silly for forgetting that the typescript magic all happens during build.

I spent entirely too much time today trying to get something, anything, to complain to me about this, to no avail. I have wrapped BaseQueryParams in a Partial.

joshuatf commented on packages/js/data/src/product-categories/types.ts:
```diff
@@ -0,0 +1,61 @@
+};
+
+type Query = BaseQueryParams< keyof ProductCategory > & {
+	context?: string;
```
I believe this one is already included in `BaseQueryParams`.

jacob-sewell replied:
True.

joshuatf commented on packages/js/data/src/product-categories/types.ts:
```diff
@@ -0,0 +1,61 @@
+
+type ProductCategoryImage = {
+	id: number;
+	date_created: string;
```
We may need to figure out a way to make nested types read only as well.

Not something that we need to solve in this PR since it won't affect the build, but would be nice to have during development.

joshuatf reviewed (commented):
Thanks for the new data store, @jacob-sewell!  Left a few comments around typing, but otherwise testing well.  🎉

Luc45 committed:
Skip test result generation and upload in forks (#34023)

* Skip test result generation and upload in forks

* Generate test report in fork

roykho committed:
Add install and build step so paths are found (#34002)

octaedro committed:
Add validations to `recordEvent` (#33911)

* Add validations to `recordEvent`

* Add changelog

* Use always class constant

* Changed trow exception by console.error

* Add tests to `validate_and_sanitize`

* Created 2 different methods to test `event_validation_and_sanitization`

* Fix tests

Co-authored-by: Fernando Marichal <contacto@fernandomarichal.com>

jonathansadowski committed:
Split changelog steps from code-freeze action (#34027)

* Split changelog script from code-freeze

* Remove hardcoded repository info from testing

* Remove generation of changelog.txt file from automation

* Correct comments and add back write of readme

albarin committed:
Update composer.lock after updating to WooCommerce Blocks 8.0.0 (#34021)

psealock committed:
Delete changelog files based on PR 34031 (#34032)

Delete changelog files for 33990

psealock committed:
Delete changelog files based on PR 34033 (#34034)

Delete changelog files for 34021

psealock committed:
Packages Release: Add manual workflow (#33922)

adrianduffell committed:
Update PHP pre-requisite description in README (#34012)

louwie17 committed:
Update woocommerce admin scripts to align more with turborepo (#33982)

* Update woocommerce admin scripts to align more with monorepo

* Add changelog

* Add dev script back in

* Add install-if-deps-outdated back into build script

* Remove --force from dev script and add WC_ADMIN_PHASE to turbo cache key

* Remove install-if-deps-outdated from build

* Add composer install step to pr unit tests workflow

* Remove old install scripts, that are not needed anymore

* Make sure client changes are part of the cache hit

* Add composer install to code coverage and code sniff as well

* Add composer_no_dev input to install-build workflow

* Add sass dependency to woocommerce/admin/client turbo cache

jacob-sewell committed:
Remove redundant context param in product categories Query (already in BaseQueryParams).

Konamiman committed:
Fix code standards in orders code, and check payment method name instead of title on get order totals (#30468)

* Fix code standards errors in class-wc-meta-box-orrder-data.php

- Sanitize all input
- Escape all output
- Verify nonce on save
- Verify that all required POST parameters are present on save
- Minor fixes (periods at end of comments, function docs, Yodas...)

* Check payment method name, not title, on add_order_item_totals_payment_method_row

* Remove unnecessary nonce verification in "save" for orders.

Also move comment to new line for readability.

* Remove unnecessary nonce check in WC_Meta_Box_Order_Data::save

 The check is already performed in the code that invokes "save".

* Add transaction id to payment method string if URL doesn't exist

* Minor fix

Co-authored-by: Jorge A. Torres <jorge.torres@automattic.com>

Konamiman committed:
Use method_exists instead of throwing in AbstractServiceProvider::reflect_class_or_callable (#33960)

* Use method_exists instead of throwing in reflect_class_or_callable

In AbstractServiceProvider::reflect_class_or_callable we were
detecting whether a 'init' method exists in the class or not via
doing 'new ReflectionMethod' and catching the exception.
This caused problems with XDebug, which is bad enough, but additionally,
throwing an exception when it's not abosultely necessary
is a bad practice in general.

* Add changelog file

ilyasfoo committed:
Fix unusable historical import (#34054)

* Add resolvers to data store

* Changelog

Konamiman committed:
Reduce the amount of terms shown in attributes page (#33962)

The attributes page displays all the terms for all the existing
product attributes, which causes a performance problem in sites
with a lot of attributes and terms.

This commit limits the number of attributes shown to 100 by default,
but it also introduces a woocommerce_max_terms_displayed_in_attributes_page
filter to customize it.

frontdevde committed:
Add WooCommerce default styles for block themes (#33518)

* Add WooCommerce default styles for block themes

* Ensure sale badge remains as is in Twenty Twenty-Two

* Style the mark element in theme colors

* Update product variations dropdowns

* address feedback

* address feedback

* fix store notice position

Co-authored-by: Luigi <gigitux@gmail.com>

jorgeatorres committed:
[COT] Implement `query()` in the COT datastore (#33834)

* Add OrdersTableQuery for querying the COT tables

* First stab at `OrdersTableDataStore::query()` method

* Silence possible PHP warning

* Add docblocks & phpcs fixes

* Introduce OrdersTableMetaQuery

* Add support for meta_query

* Add changelog

* ‘status’ should have the ‘wc-‘ prefix

* PHPCS fixes

* Minor improvements

* Add tests for OrdersTableQuery

* Remove ‘paginate’ query var

* Add support for reading multiple objects at once

* Address feedback

* Prevent `get_order_data_for_ids()` from producing an incorrect query when no IDs are passed

* Make CI happy

MrJnrman committed:
Add JSON report output to PW tests (#34057)

* Add PW json report when in CI

* Added wait for flaky test

* Used more robust selectors

* Added changelog

psealock committed:
Delete changelog files based on PR 34061 (#34062)

Delete changelog files for 34054

jorgeatorres committed:
[COT] `query()`: add support for the `customer` query var (#34059)

* Make sure orders are saved with a default status

* Add support for ‘customer’ query var in COT

* Make PHPCS happy in COT datastore tests

* Add tests for ‘customer’ query var

* Add changelog

Commit:
E2E: Update WSL readme (#34068)

* Update link to setting default Linux distro in WSL

* Add changelog

jacob-sewell committed:
Apply Partial to BaseQueryParams.

ilyasfoo committed:
Fix missing product task add product track via product templates (#34070)

* Add missing track on adding products via product templates

* Changelog

Konamiman committed:
Fix: product attributes lookup table usage enabled on each upgrade (#34072)

* Stop re-enabling product attributes lookup table usage after each install.

Now the option will be enabled after a WooCommerce install/update
only if the user hadn't explicitly disabled it before.

* Add changelog file

joshuatf committed:
Add utils for testing tracks in backend (#34010)

* Add methods to get and clear queued pixel events

* Add utils for getting and asserting events in unit tests

* Add in some assertions around task list completion tracks

* Add changelog entry

* Add comment for test task fixture

joshuatf committed:
Log tracks events after they have been validated (#34030)

* Log events after they have been validated

* Add changelog entry

* Remove old tracks file

* Fix lint errors

jonathansadowski committed:
Update code-freeze action to exit earlier without fail on non-release days (#34046)

* Update code-freeze action to exit earlier without fail on non-release days

* Update freeze status to mirror exit code

MrJnrman committed:
Fix app root fetching when using PNPM version 7 (#33929)

* Updated to pass lint checks

* Used __dirname to determine package path

chihsuan committed:
Fix shipping zone fails to create after installing Jetpack and WCS (#34074)

* Fix shipping zone fails to create after installing jetpack and wcs

* Add changelog

psealock committed:
Delete changelog files based on PR 34087 (#34088)

Delete changelog files for 34070

psealock committed:
Delete changelog files based on PR 34089 (#34090)

Delete changelog files for 34072

psealock committed:
Delete changelog files based on PR 34091 (#34092)

Delete changelog files for 34074

Commit:
Use the default payment suggestions when woocommerce_show_marketplace_suggestions is set to no (#34083)

* Support force_default_suggestions argument

* Use force_default_suggestions

* Support force_default_suggestions argument -- returns the default payment suggestions when both woocommerce_show_marketplace_suggestions and woocommerce_setting_payments_recommendations_hidden options are set to no

* Add changelog

adrianduffell committed:
Fix Tracks event in Review Shipping task (#34045)

psealock committed:
Delete changelog files based on PR 34093 (#34094)

Delete changelog files for 34045

ilyasfoo committed:
Fix shipping zone count transient reset (#34099)

* Delete shipping zone transient on adding shipping zone via smart defaults

* Changelog

* Fix comment

vedanshujain committed:
Render order edit page  (#33638)

* Implement order edit screen rendering.

* Add changelog.

* Use global function instead of harcoded string.

* Use pagecontroller to display order edit form.

* Move HTML out from the translatable string.

Co-authored-by: Barry Hughes <3594411+barryhughes@users.noreply.github.com>

* Capitalize `Order` for consistency.

Co-authored-by: Barry Hughes <3594411+barryhughes@users.noreply.github.com>

* Minor realignment.

* phpcs fix .

* Decouple meta boxes from global $post object to support custom order tables.

* Bug fixes - improperly set $order object, and allow passing null.

* Make order item meta work with order object.

* Use `$order` object to minimize diff.

* Refactor $theorder object initialization to a common method.

* Remove check for WC_Order object.

* Move common methods to inside `src` in a dedicated class.

* Fix spacing for code style.

* Renamed method to be more accurate.

* Add proxies for static so they can be mocked.

* Remove PageController import.

* Move helper methods to an internal util class.

Co-authored-by: Barry Hughes <3594411+barryhughes@users.noreply.github.com>

Konamiman committed:
Introduce the object cache infrastructure (#33634)

* Introduce the ObjectCache class.

It's an abstract class intended to be used to implement caching for
any kind of object (arrays or PHP objects) using the underlying
WordPress caching or any alternative custom cache engine.

* Add changelog file

* Change the order of parameters in ObjectCache::get

Also change the max expiration value from one day to one month

daniyalahmadk committed:
Fix fatal error when bulk editing sale price to empty value in PHP 8 (#34053)

Commit:
Clean up setup_experiment_1 and setup_experiment_2 experiments (#34084)

* Remove setup_experiment_1 and setup_experiment_2 tests and css

* Remove setup_experiment_1 and 2 from the client code

* Copy over setup_experiment_1 options to the setup tasklist and remove setup_experiment_1 and 2

* Add changelog

* Determine the default component to use by task id

* Update plugins/woocommerce-admin/client/tasks/hooks/useActiveSetupList.ts

Co-authored-by: Chi-Hsuan Huang <chihsuan.tw@gmail.com>

Co-authored-by: Chi-Hsuan Huang <chihsuan.tw@gmail.com>

psealock committed:
Beta Tester: Update composer lock file (#34086)

* Update Beta Tester lock file

* changelog

psealock committed:
Delete changelog files based on PR 34116 (#34117)

Delete changelog files for 34099

rjchow committed:
fix: wrong condition for shipping recommendations tour step (#34120)

barryhughes committed:
Column sorting for the COT admin list table. (#33787)

* Column sorting for the COT admin list table.

* Add filter for easier extensibility.

* Changelog.

* Add support for orderby order’s total and creation/modification dates

* Remove unnecessary import

* Use ‘any’ instead of ‘all’ as status so that all valid statuses are used instead

* Use ‘order_total’ as orderby GET arg for backwards compat

* Rework sorting logic after support in `OrdersTableQuery`

* Update `@since` tag

* Make PHPCS happy

Co-authored-by: Jorge A. Torres <jorge.torres@automattic.com>

joshuatf reviewed (dismissed):
Testing wonderfully and typing looks great!  Also appreciate the extensive documentation here.  Nice work, @jacob-sewell!

jacob-sewell committed:
Merge branch 'trunk' into 33464/add-product-categories-datastore-api

joshuatf reviewed (approved):
Testing well for me still!  :shipit:
